flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -96,52 +97,88 @@ public DefaultExecutionSlotAllocator(
 
 			LOG.debug("Allocate slot with id {} for execution {}", slotRequestId, executionVertexId);
 
-			CompletableFuture<LogicalSlot> slotFuture = calculatePreferredLocations(
+			final CompletableFuture<SlotProfile> slotProfileFuture = createSlotProfile(
+				schedulingRequirements,
+				Collections.emptySet(),
+				schedulingRequirements.getPhysicalSlotResourceProfile(),
+				allPreviousAllocationIds);
+
+			final CompletableFuture<LogicalSlot> slotFuture = slotProfileFuture.thenCompose(
+				slotProfile ->
+					slotProviderStrategy.allocateSlot(
+						slotRequestId,
+						new ScheduledUnit(
+							executionVertexId,
+							slotSharingGroupId,
+							schedulingRequirements.getCoLocationConstraint()),
+						slotProfile));
+
+			final SlotExecutionVertexAssignment slotExecutionVertexAssignment =
+				createAndRegisterSlotExecutionVertexAssignment(
 					executionVertexId,
-					schedulingRequirements.getPreferredLocations(),
-					inputsLocationsRetriever,
-					Collections.emptySet()).thenCompose(
-							(Collection<TaskManagerLocation> preferredLocations) ->
-								slotProviderStrategy.allocateSlot(
-									slotRequestId,
-									new ScheduledUnit(
-										executionVertexId,
-										slotSharingGroupId,
-										schedulingRequirements.getCoLocationConstraint()),
-									SlotProfile.priorAllocation(
-										schedulingRequirements.getTaskResourceProfile(),
-										schedulingRequirements.getPhysicalSlotResourceProfile(),
-										preferredLocations,
-										Collections.singletonList(schedulingRequirements.getPreviousAllocationId()),
-										allPreviousAllocationIds)));
-
-			SlotExecutionVertexAssignment slotExecutionVertexAssignment =
-					new SlotExecutionVertexAssignment(executionVertexId, slotFuture);
-			// add to map first to avoid the future completed before added.
-			pendingSlotAssignments.put(executionVertexId, slotExecutionVertexAssignment);
-
-			slotFuture.whenComplete(
-					(ignored, throwable) -> {
-						pendingSlotAssignments.remove(executionVertexId);
-						if (throwable != null) {
-							slotProviderStrategy.cancelSlotRequest(slotRequestId, slotSharingGroupId, throwable);
-						}
-					});
+					slotFuture,
+					slotRequestId,
+					slotSharingGroupId);
 
 			slotExecutionVertexAssignments.add(slotExecutionVertexAssignment);
 		}
 
 		return slotExecutionVertexAssignments;
 	}
 
-	private void validateSchedulingRequirements(Collection<ExecutionVertexSchedulingRequirements> schedulingRequirements) {
+	protected void validateSchedulingRequirements(Collection<ExecutionVertexSchedulingRequirements> schedulingRequirements) {
 		schedulingRequirements.stream()
 			.map(ExecutionVertexSchedulingRequirements::getExecutionVertexId)
 			.forEach(id -> checkState(
 				!pendingSlotAssignments.containsKey(id),
 				"BUG: vertex %s tries to allocate a slot when its previous slot request is still pending", id));
 	}
 
+	protected CompletableFuture<SlotProfile> createSlotProfile(
```

Comment:
This is a shared method. Not sure what is the duplication?

Reply:
True, there is no duplication at the moment.
I meant to move the concern of `SlotProfileRetriever` (and all related code) into a separate component to reuse it in `OneSlotPerExecutionSlotAllocator/DefaultExecutionSlotAllocator`. This would simplify `OneSlotPerExecutionSlotAllocator/DefaultExecutionSlotAllocator`, imo, and there will be also no code duplication.

Reply:
In the latest update, I removed `createSlotProfile()` because I find it does not reduce much code duplication but introduced a nested method with weird parameters to distinguish `DefaultExecutionSlotAllocator` and `OneSlotPerExecutionSlotAllocator `.

I think the major part that would be good to extract out from `ExecutionSlotAllocator` is the preferred location calculation. We can introduce a `PreferredLocationsRetriever` which wraps `InputsLocationsRetriever`, `calculatePreferredLocations(...)` and `getPreferredLocationsBasedOnInputs(...)`. 
I think we can open another task to do it since it is not very related  to this change. (I'm already working on it though.)
WDYT?

Reply:
FLINK-18034 is opened to "Introduce PreferredLocationsRetriever".
With it we can go one step further to centrally managed the preferred location calculation regardless of state-based or inputs-based.

Reply:
PR #12423 is opened which is based on this PR.